ksmt-yices/src/main/kotlin/org/ksmt/solver/yices/KYicesExprConverter.kt

```diff
@@ -0,0 +1,631 @@
+        }
+    }
+
+    private fun convertAtomic(expr: YicesTerm) = with(ctx) {
+        when (Terms.constructor(expr)) {
+            Constructor.BOOL_CONSTANT -> convert { Terms.boolConstValue(expr).expr }
+            Constructor.ARITH_CONSTANT -> convert {
+                val value = Terms.arithConstValue(expr)
+
+                mkRealNum(mkIntNum(value.numerator), mkIntNum(value.denominator))
+            }
+
+            Constructor.BV_CONSTANT -> convert {
+                val sizeBits = Terms.bitSize(expr)
+                val value = Terms.bvConstValue(expr).map { if (it) 1 else 0 }.reversed().joinToString(separator = "")
+
+                mkBv(value, sizeBits.toUInt())
+            }
+
+            Constructor.VARIABLE, Constructor.UNINTERPRETED_TERM -> convert {
+                val convertedDecl = convertDecl(expr)
+                check(convertedDecl is KConstDecl<*>) { "Unexpected declaration $convertedDecl" }
+
+                convertedDecl.apply()
+            }
+            else -> error("Not supported term ${Terms.toString(expr)}")
+        }
+    }
+
+    private fun convertProjection(expr: YicesTerm) = with(ctx) {
+        val constructor: Constructor = Terms.constructor(expr)
+        check(constructor == Constructor.BIT_TERM) { "Not supported term ${Terms.toString(expr)}" }
+
+        expr.convert(arrayOf(Terms.projArg(expr))) { bv: KExpr<KBvSort> ->
+            val bv1 = mkBvExtractExpr(Terms.projIndex(expr), Terms.projIndex(expr), bv)
+            val condition = mkEq(mkBv(true, 1u), bv1)
+            mkIte(condition, mkTrue(), mkFalse())
+        }
+    }
+
+    private fun convertSum(expr: YicesTerm) = with(ctx) {
+        val (consts, children) = List(Terms.numChildren(expr)) { index ->
+            val component = Terms.sumComponent(expr, index)
+
+            Pair(
+                mkRealNum(mkIntNum(component.constValue.numerator), mkIntNum(component.constValue.denominator)),
+                component.term
+            )
+        }.unzip().let { (l1, l2) -> Pair(l1.toMutableList(), l2.toMutableList()) }
+
+        val freeConst = if (children.first() == Terms.NULL_TERM) {
+            children.removeFirst()
+            consts.removeFirst()
+        } else {
+            mkRealNum(0)
+        }
+
+        argTypes.addAll(List(children.size) { realSort })
+
+        expr.convertList(children.toTypedArray()) { args: List<KExpr<KRealSort>> ->
+            args.zip(consts)
+                .map { mkArithMul(it.first, it.second) }
+                .fold(freeConst) { acc: KExpr<KRealSort>, t: KExpr<KRealSort> ->
+                    mkArithAdd(acc, t)
+                }
+        }
+    }
+
+    private fun convertBvSum(expr: YicesTerm) = with(ctx) {
+        val bvSize = Terms.bitSize(expr).toUInt()
+        val (consts, children) = List(Terms.numChildren(expr)) { index ->
+            val component = Terms.sumbvComponent(expr, index)
+            val bvConstValue = component.constValue
```

Comment:
It is better to define some function for `bool[] -> bv` and use it here and elsewhere.